### PR TITLE
Update Nifi to 1.14.0 version

### PR DESCRIPTION
**Pull request by zakaria2905**
below, the main changes brought by this PR:

- Nifi has been updated to version 1.14.0. This version comes already secured.
- Adding of differents ways to log in Nifi (Single-User, OIDC, LDAP)
- Adding of OpenLDAP dependencie 
- Adapting files in conf/ to provide OIDC and LDAP
- Adding doc folder to show how to configure a Single-User, OIDC or LDAP authentication
- Modifying statefulset to fix bugs

### CODE REVIEW

File: values.yaml [hunk 1]
```diff
@@ -69,13 +69,13 @@ sts:
 
 properties:
   # use externalSecure for when inbound SSL is provided by nginx-ingress or other external mechanism
+  sensitiveKey: changeMe # Must to have minnimal 12 length key
   externalSecure: false
-  isNode: true # set to false if ldap is enabled
-  httpPort: 8080 # set to null if ldap is enabled
-  httpsPort: null # set to 9443 if ldap is enabled
-  webProxyHost:
+  isNode: flase
```
Comment (drivard) [inline values.yaml:74]: typo in `false`

**Comment by Jarzamendia** [comment]
This feature is great!

File: configs/nifi.properties [hunk 2]
```diff
@@ -147,7 +147,7 @@ nifi.web.jetty.threads=200
 # nifi.web.proxy.context.path=
 
 # security properties #
-nifi.sensitive.props.key=
+nifi.sensitive.props.key={{ .Values.properties.sensitiveKey}}
 nifi.sensitive.props.key.protected=
 nifi.sensitive.props.algorithm=PBEWITHMD5AND256BITAES-CBC-OPENSSL
```
Comment (drivard) [inline configs/nifi.properties:152]: I tested out these changes. I am on a brand new install of 1.14.0. I faced the following issue: https://www.mail-archive.com/users@nifi.apache.org/msg14466.html 

```
> Hi,
> according to NIFI-8246 defaults are
> 1.14.0:
> nifi.sensitive.props.algorithm=NIFI_PBKDF2_AES_GCM_256
> 1.13.2
> nifi.sensitive.props.algorithm=PBEWITHMD5AND256BITAES-CBC-OPENSSL
>
> The same passwords don't work
>
> My workaround was to export template from an old NiFi and to import it to
> a new one.
> Hope it helps )
>
> Alexey
```

So to get 1.14.0 to boot on a brand new install I had to change the setting to `nifi.sensitive.props.algorithm=NIFI_PBKDF2_AES_GCM_256` I don't know how nifi is going to react on an instance with previous data.

Comment (zakaria2905) [reply configs/nifi.properties:152]: Hi Drivard,

Thank you for these information. I will make these changes in the next update.

File: .helmignore [hunk 3]
```diff
@@ -24,3 +24,6 @@
 *.tmproj
 
 .circleci/
+
+.doc/
```
Comment (drivard) [inline .helmignore:28]: it doesn't trigger the right folder. It should be just `doc/` this cause the error `Error: INSTALLATION FAILED: create: failed to create: Request entity too large: limit is 3145728`. thanks

Comment (zakaria2905) [reply .helmignore:28]: Thank you for your contribution (again)

**Comment by drivard** [comment]
Just a heads up, while I am working on a nifi  1.14.0 secured cluster with oidc login using keycloak like you do. I finally got somewhere. But I had to add multiple fixes to the chart, here is the list:

`authorizers.xml` Adding the OIDC initial admin user and the node as user identity because otherwise they are not able to create the cluster.
image

`authorizers.xml` Modifying the access policy provider
image

`values.yaml` adding an oidc admin key under the oidc section which will be used for the initial user login

`nifi.properties` added the identity properties, plus the sensitive key and algorithm which has been talk here already
image

`statefulset.yaml` had to remove the -t in the timeout command line
image

`service.yaml` add to fix the service nodeport
image

`ingress.yaml` add to set the ingress (nginx) version to latest version
image

Not showing the email logged in, but the cluster is live with 3 node:
image

**Comment by drivard** [comment]
for the nginx ingress to work I had to add a few annotations but there is still a small issue on the sticky session.

```kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
    nginx.ingress.kubernetes.io/ssl-passthrough: "true"
    nginx.ingress.kubernetes.io/affinity-mode: persistent
    nginx.ingress.kubernetes.io/affinity: "cookie"```

File: values.yaml [hunk 4]
```diff
@@ -69,18 +69,17 @@ sts:
...
-  clusterSecure: false # set to true if ldap is enabled
+  clusterSecure: true
   needClientAuth: false
   provenanceStorage: "8 GB"
   siteToSite:
     port: 10000
-  authorizer: managed-authorizer
   # use properties.safetyValve to pass explicit 'key: value' pairs that overwrite other configuration
   safetyValve:
     #nifi.variable.registry.properties: "${NIFI_HOME}/example1.properties, ${NIFI_HOME}/example2.properties"
```
Comment (drivard) [inline values.yaml:85]: if you want the port-forward to workout, I figured out this morning that you need to setup the following https properties under safetyValve:

```
safetyValve:
    #nifi.variable.registry.properties: "${NIFI_HOME}/example1.properties, ${NIFI_HOME}/example2.properties"
    nifi.web.http.network.interface.default: eth0
    # listen to loopback interface so "kubectl port-forward ..." works
    nifi.web.http.network.interface.lo: lo
    # Cluster Secure requirements
    nifi.web.https.network.interface.default: eth0
    # listen to loopback interface so "kubectl port-forward ..." works
    nifi.web.https.network.interface.lo: lo
```

**Comment by bmgante** [comment]
Hi, what are the plans/expectations to release this?

**Comment by bmgante** [comment]
Ok thanks for clarification. Btw, I noticed that current 1.12.1 has some issues regarding content repository cleanup even having archiving disabled or content in volatile mode. For some reason, content is not properly deleted and easily reaches the max storage. In the meantime I raised this thread in slack and someone is pointing to some issue that seem to have been added in 1.12 release and to be fixed only in 1.15.

https://apachenifi.slack.com/archives/C0L9VCD47/p1634593368451400

My questions are:
- any of you are facing this same issue?
- if confirmed it's a bug and as I can't wait for 1.15 I am wondering to downgrade to 1.11. In this case, how could I use this helm chat to use nifi 1.11? What changes would I need to perform in the helm chart?

Thanks in advance

File: configs/nifi.properties [hunk 5]
```diff
@@ -147,9 +147,9 @@ nifi.web.jetty.threads=200
 # nifi.web.proxy.context.path=
 
 # security properties #
-nifi.sensitive.props.key=
+nifi.sensitive.props.key={{.Values.properties.sensitiveKey}}
```
Comment (wknickless) [inline configs/nifi.properties:150]: Not comfortable putting the sensitive key in a nifi-values.yaml file that would be (e.g.) checked into a repository; see my pull request for an alternative approach.

File: configs/nifi.properties [hunk 6]
```diff
@@ -147,9 +147,9 @@ nifi.web.jetty.threads=200
...
 nifi.sensitive.props.key.protected=
-nifi.sensitive.props.algorithm=PBEWITHMD5AND256BITAES-CBC-OPENSSL
+nifi.sensitive.props.algorithm=NIFI_PBKDF2_AES_GCM_256
```
Comment (wknickless) [inline configs/nifi.properties:152]: If you're changing a default, would you please make it configurable so we can continue to use the old default if we already have a bunch of things protected with a key?

**Comment by thanhhaiduong** [comment]
> Just a heads up, while I am working on a nifi 1.14.0 secured cluster with oidc login using keycloak like you do. I finally got somewhere. But I had to add multiple fixes to the chart, here is the list:
> 
> `authorizers.xml` Adding the OIDC initial admin user and the node as user identity because otherwise they are not able to create the cluster. image
> 
> `authorizers.xml` Modifying the access policy provider image
> 
> `values.yaml` adding an oidc admin key under the oidc section which will be used for the initial user login
> 
> `nifi.properties` added the identity properties, plus the sensitive key and algorithm which has been talk here already image
> 
> `statefulset.yaml` had to remove the -t in the timeout command line image
> 
> `service.yaml` add to fix the service nodeport image
> 
> `ingress.yaml` add to set the ingress (nginx) version to latest version image
> 
> Not showing the email logged in, but the cluster is live with 3 node: image
> 
> thanks to this [terraform](https://github.com/pvillard31/nifi-gcp-terraform/blob/master/gcp-cluster-secured-nifi-oidc/nifi.tf) recipe which helped me figure out the last bit using the Node Identity to stop getting the Untrusted proxy error message.

can you help me? I can't setup cluster with helm charts.

**Comment by drivard** [comment]
@banzo @zakaria2905 have you tried since this merge to create a cluster of 3 nodes?

Before I create an issue it seems to me that since the central CA option has been removed from the chart 

(https://github.com/cetic/helm-nifi/pull/169/files#diff-1d048f4995b53d97babf0d21a6c76ad5a14c3cfe0b9b1a79a81ed08366967955L105-L160)

it looks like when you create a cluster of 3 nodes they all considers themselves as just single nodes not being part of a cluster. 

Maybe we need a bit of help or documentation on how the clustering works now with a standalone only tls-toolkit.sh

**Comment by thanhhaiduong** [comment]
> how the clustering works now with a standalone only tls-toolkit.sh

Can I setup cluster authen with singleuser with helm chart?
can you help me?

**Comment by drivard** [comment]
I was unable to set a cluster using the latest version, I have passed several hours trying to re-integrate the CA chart but I let it go.

The only way I was able to do start a cluster, was to checkout the code before this merged PR and by adding all the described changes in the code I wrote in my previous comment. 

I am running a 1.15.0 cluster as we speak using the old way with the CA chart.

**Comment by thanhhaiduong** [comment]
> I was unable to set a cluster using the latest version, I have passed several hours trying to re-integrate the CA chart but I let it go.
> 
> The only way I was able to do start a cluster, was to checkout the code before this merged PR and by adding all the described changes in the code I wrote in my previous comment.
> 
> I am running a 1.15.0 cluster as we speak using the old way with the CA chart.

can you give me your helm chart that you have successfully implemented with cluster?

**Comment by drivard** [comment]
Sorry way too much customized.

**Comment by thanhhaiduong** [comment]
> Sorry way too much customized.

I have setup cluster with helm chart, and I get error:
This node is currently not connected to the cluster. Any modifications to the data flow made here will not replicate across the cluster.
image

My logs: 
[Process Cluster Protocol Request-18] o.a.n.c.p.impl.SocketProtocolListener Failed processing protocol message from nifi-1.nifi-headless.nifi.svc.cluster.local due to Received fatal alert: certificate_unknown

Can you help me?

**Pull request merged**